homeassistant/components/brunt/cover.py

```diff
@@ -1,69 +1,83 @@
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .const import (
+    ATTR_REQUEST_POSITION,
+    ATTRIBUTION,
+    CLOSED_POSITION,
+    DOMAIN,
+    OPEN_POSITION,
+)
 
 _LOGGER = logging.getLogger(__name__)
 
 COVER_FEATURES = SUPPORT_OPEN | SUPPORT_CLOSE | SUPPORT_SET_POSITION
-
-ATTR_REQUEST_POSITION = "request_position"
-NOTIFICATION_ID = "brunt_notification"
-NOTIFICATION_TITLE = "Brunt Cover Setup"
-ATTRIBUTION = "Based on an unofficial Brunt SDK."
-
-CLOSED_POSITION = 0
-OPEN_POSITION = 100
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_USERNAME): cv.string, vol.Required(CONF_PASSWORD): cv.string}
-)
+REGULAR_INTERVAL = timedelta(seconds=20)
+FAST_INTERVAL = timedelta(seconds=1)
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_entry(hass, entry, async_add_entities, discovery_info=None):
     """Set up the brunt platform."""
+    bapi: BruntClientAsync = hass.data[DOMAIN][entry.entry_id]
 
-    username = config[CONF_USERNAME]
-    password = config[CONF_PASSWORD]
-
-    bapi = BruntAPI(username=username, password=password)
-    try:
-        things = bapi.getThings()["things"]
-        if not things:
-            _LOGGER.error("No things present in account")
-        else:
-            add_entities(
-                [
-                    BruntDevice(bapi, thing["NAME"], thing["thingUri"])
-                    for thing in things
-                ],
-                True,
-            )
-    except (TypeError, KeyError, NameError, ValueError) as ex:
-        _LOGGER.error("%s", ex)
-        hass.components.persistent_notification.create(
-            "Error: {ex}<br />You will need to restart hass after fixing.",
-            title=NOTIFICATION_TITLE,
-            notification_id=NOTIFICATION_ID,
-        )
-
-
-class BruntDevice(CoverEntity):
+    async def async_update_data():
+        """Fetch data from the Brunt endpoint for all Things."""
+        try:
+            async with async_timeout.timeout(10):
+                things = await bapi.async_get_things(force=True)
+                return {thing.SERIAL: thing for thing in things}
+        except ServerDisconnectedError as err:
+            raise UpdateFailed(f"Error communicating with API: {err}") from err
+        except ClientResponseError as err:
+            if err.status == 403:
+                raise ConfigEntryAuthFailed() from err
+            if err.status == 401:
+                raise UpdateFailed(f"Error communicating with API: {err}") from err
+        except Exception as err:
+            raise UpdateFailed("Unknown error.") from err
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="brunt",
+        update_method=async_update_data,
+        update_interval=REGULAR_INTERVAL,
+    )
+    await coordinator.async_config_entry_first_refresh()
```

Comment:
This coordinator and setup login belong in the init file when setting up the entry. Most logic is on that end already. Just pass the coordinator around via hass.data.

Additionally, the `async_config_entry_first_refresh` only has use for the entry setup